File: xmlBlaster/src/java/org/xmlBlaster/util/cluster/NodeId.java
/*------------------------------------------------------------------------------
Name:      NodeId.java
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
Comment:   Holds the unique name of a cluster node
Author:    xmlBlaster@marcelruff.info
------------------------------------------------------------------------------*/
package org.xmlBlaster.util.cluster;

import java.util.logging.Logger;


/**
 * Holds the unique name of an xmlBlaster server instance (= cluster node)
 * @author xmlBlaster@marcelruff.info 
 * @since 0.79e
 * @see <a href="http://www.xmlBlaster.org/xmlBlaster/doc/requirements/cluster.html">cluster requirement</a>
 */
public final class NodeId implements Comparable, java.io.Serializable
{
   private static final long serialVersionUID = 1L;
   private static Logger log = Logger.getLogger(NodeId.class.getName());
   private static final String ME = "NodeId";

   /**
    * @see #setId(String)
    */
   public NodeId(String id) {
      setId(id);
   }

   /**
    * @return e.g. "heron" (/node/heron/... is stripped)
    */
   public final String getId() {
      return id;
   }

   /**
    * You need to pass a valid node name without any special characters,
    * e.g. "http://xy.com:3412" is not allowed as it contains '/' and ':' chars. 
    * @param id The cluster node id, e.g. "heron".<br />
    * If you pass "/node/heron/client/joe" everything is stripped to get "heron"
    * @see org.xmlBlaster.util.Global#getStrippedId()
    */
   public final void setId(String id) {
      if (id == null || id.length() < 1) {
         log.severe("Cluster node has no name please specify one with -cluster.node.id XXX");
         id = "NoNameNode";
      }
      this.id = id;
      if (this.id.startsWith("/node/"))
         this.id = this.id.substring("/node/".length()); // strip leading "/node/"

      int index = this.id.indexOf("/");   // strip tailing tokens, e.g. from "heron/client/joe" make a "heron"
      if (index == 0) {
         throw new IllegalArgumentException(ME+": The given cluster node ID '" + id + "' may not start with a '/'");
      }
      if (index > 0) {
         this.id = this.id.substring(0, index);
      }
   }

   /**
    * @see #getId()
    */
   public final String toString() {
      return getId();
   }
 
   /**
    * Needed for use in TreeSet and TreeMap, enforced by java.lang.Comparable
    */
   public final int compareTo(Object obj)  {
      NodeId n = (NodeId)obj;
      return toString().compareTo(n.toString());
   }

   public final boolean equals(NodeId n) {
      if (id == null) {
         if (n == null || n.getId() == null) return true;
         return false;
      }
      if (n == null || n.getId() == null) return false;
      return id.equals(n.getId());
   }
  
   private String id;
}

File: src/c++/util/cluster/NodeId.cpp
/*------------------------------------------------------------------------------
Name:      NodeId.cpp
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
Comment:   Holds the unique name of a cluster node
------------------------------------------------------------------------------*/


/**
 * Holds the unique name of an xmlBlaster server instance (= cluster node)
 * @author xmlBlaster@marcelruff.info 
 * @author laghi@swissinfo.org
 * @since 0.79e
 * @url http://www.xmlBlaster.org/xmlBlaster/doc/requirements/cluster.html
 */

#include <util/cluster/NodeId.h>
#include <util/Global.h>
#include <util/lexical_cast.h>

namespace org { namespace xmlBlaster { namespace util { namespace cluster {

using namespace std;
using namespace org::xmlBlaster::util;

NodeId::NodeId(Global& global, const string& id)
: ME("NodeId"), global_(global), log_(global.getLog("org.xmlBlaster.cluster"))
{
   setId(id);
}

NodeId::NodeId(const NodeId& nodeId) : ME("NodeId"), global_(nodeId.global_), log_(nodeId.log_)
{
   setId(nodeId.id_);
}

NodeId& NodeId::operator =(const NodeId& nodeId)
{
   setId(nodeId.id_);
   return *this;
}

string NodeId::getId() const
{
   return id_;
}

/**
 * @param id The cluster node id, e.g. "heron".<br />
 * If you pass "/node/heron/client/joe" everything ins stripped to get "heron"
 */
void NodeId::setId(const string& id)
{
   if (id.empty()) {
//      log_.error(ME, "Cluster node has no name");
      id_ = "NoNameNode";
   }
   id_ = id;

   if (id_.find_first_of("/node/") == 0)
      id_ = id_.substr(string("/node/").length()); // strip leading "/node/"
   string::size_type index = id_.find_first_of("/");   // strip tailing tokens, e.g. from "heron/client/joe" make a "heron"
   if (index == 0) {
      throw XmlBlasterException(INTERNAL_ILLEGALARGUMENT, ME, "setId: The given cluster node ID '" + lexical_cast<std::string>(id_) + "' may not start with a '/'");
   }
   if (index > 0) {
      id_ = id_.substr(0, index);
   }
}

string NodeId::toString() const
{
   return getId();
}
 
/**
 * Needed for use in TreeSet and TreeMap, enforced by java.lang.Comparable
 */
bool NodeId::operator <(const NodeId& nodeId) const
{
   return toString() < nodeId.toString();
}

bool NodeId::operator ==(const NodeId& nodeId) const
{
   return toString() == nodeId.toString();
}

}}}} // namespace
